docroot/modules/custom/mass_entity_usage/mass_entity_usage.module

```diff
@@ -0,0 +1,116 @@
+<?php
+
+/**
+ * @file
+ * Contains mass_entity_usage.module.
+ */
+
+use Drupal\content_moderation\Entity\ContentModerationState;
+use Drupal\content_moderation\Entity\ContentModerationStateInterface;
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\mass_content_moderation\MassModeration;
+use Drupal\mayflower\Helper;
+
+/**
+ * Implements hook_entity_usage_block_tracking().
+ */
+function mass_entity_usage_entity_usage_block_tracking($target_id, $target_type, $source_id, $source_type, $source_langcode, $source_vid, $method, $field_name, $count) {
+  // Block tracking of entities with a node source that is unpublished or trash.
+  if (!_mass_entity_usage_tracking_check($source_type, $source_id, $source_vid)) {
+    return TRUE;
+  }
+  return FALSE;
+}
+
+/**
+ * Find an id/revision_id record in the default field data table.
+ *
+ * @param string $entity_type
+ *   The entity type.
+ * @param int $id
+ *   The entity id.
+ * @param int $vid
+ *   The entity revision id.
+ *
+ * @return bool
+ *   If true, the content should be tracked. If false, do not track.
+ */
+function _mass_entity_usage_tracking_check($entity_type, $id, $vid) {
+  if ($entity_type == 'paragraph') {
+    // Verify the paragraph revision is in the paragraphs_item_field_data table.
+    $query = \Drupal::database()->select('paragraphs_item_field_data', 'pifd');
```

Comment:
Have we checked the performance if this query with EXPLAIN? Could cause slowness if there is no index.

Reply:
I haven't. But it makes sense to create an index on id and revision_id on the paragraphs_item_field_data, as insertions will be fewer compared to the number of reads.
I will create an index for that. Thanks!